galley/pkg/mcp/server/server.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"sync/atomic"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/protobuf/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"istio.io/istio/pkg/log"
+)
+
+var (
+	mcpLog = log.RegisterScope("mcp", "mcp debugging", 0)
+)
+
+// WatchResponse contains a versioned collection of pre-serialized resources.
+type WatchResponse struct {
+	TypeURL string
+
+	// Version of the resources in the response for the given
+	// type. The client responses with this version in subsequent
+	// requests as an acknowledgment.
+	Version string
+
+	// Responses to be included in the response.
+	Resources []Resource
+}
+
+// CancelWatchFunc allows the consumer to cancel a previous watch,
+// terminating the watch for the request.
+type CancelWatchFunc func()
+
+// Watcher requests watches for configuration resources by node, last
+// applied version, and type. The watch should send the responses when
+// they are ready. The watch can be canceled by the consumer.
+type Watcher interface {
+	// CreateWatch returns a new open watch for a non-empty request.
+	//
+	// Immediate responses should be returned to the caller along
+	// with an optional cancel function. Asynchronous responses should
+	// be delivered through the write-only WatchResponse channel. If the
+	// channel is closed prior to cancellation of the watch, an
+	// unrecoverable error has occurred in the producer, and the consumer
+	// should close the corresponding stream.
+	//
+	// Cancel is an optional function to release resources into the
+	// producer. If provided, the consume may call this function
+	// multiple times.
```

Comment:
```If provided, the consume may call this function multiple times?```
Perhaps a different way to phrase it would be: 
It can be called idempotently to cancel and release resources?